### ISSUE TITLE
Reduce build duration by lowering Mockito usage

### ISSUE THREAD

**Issue by jkronegg** [cucumber/cucumber-jvm#2707]
### 🤔 What's the problem you've observed?

Mockito has a impact on the unit test execution time.

|     project     |  test duration [s] |      mockito contribution      | 
| --------------- | ------------------ | ------------------------------ | 
| cucumber-junit  |                4.4 |                    31% = 1.3 s |
| cucumber-core   |                7.5 |                    10% = 0.7 s |
| cucumber-java   |                2.8 |                    24% = 0.6 s |
| cucumber-java8  |                1.8 |                    27% = 0.5 s |
| cucumber-spring |                4.5 |                    13% = 0.5 s |
| cucumber-testng |                1.5 |                    20% = 0.3 s |
| cucumber-guice  |                1.8 |                    15% = 0.3 s |
| datatable       |                0.5 |   (unused dependency) 0% = 0 s |

Total build duration: 3min17 = 197 seconds
Total test duration : 23.5 seconds
Total Mockito contribution : 4.2 seconds (18% of tests, 2% of build duration)

By reducing the usage of Mockito, the test duration can be lowered with two advantages:
- this gives the developer a shorter feedback loop.
- the energy consumption is lower (green IT), which is important because the cucumber-jvm project is built often (due to PR and renovate bot). 2% is not a lot, but on the long term, it makes a difference.

### ✨ Do you have a proposal for making it better?

I didn't check in details, but for example in `JavaBackendTest`, the `@Mock` `ObjectFactory` and `Glue` could be replaced by fake implementations, which could reduce the class test duration by about 80%.

**Comment by mpkorstanje** [comment]
Yes. This makes sense. I would agree to Mockito replacements if:

1. Initially we keep the stubs in the test class.
2. Keep the stubs simple.

This will likely result in some duplication but that is an acceptable tradeoff.

**Comment by CharlesLgn** [comment]
Does someone made a study to use another  mock library to see if there will be better performance ? (EasyMock/JMockit/...)

**Comment by jkronegg** [comment]
I didn't tried with other mock library. The proposed solution #2767 replaces mockito by stub and fake implementations, which leads to a 7% improvement of the build time (that is, much more than the 2% expected).

**Comment by CharlesLgn** [comment]
> I didn't tried with other mock library. The proposed solution #2767 replaces mockito by stub and fake implementations, which leads to a 7% improvement of the build time (that is, much more than the 2% expected).

I understand the principle of using fake implementation. However, I think that if you have an interface with 25 methods, the test class will be poluate with a lot of fake implementation, with also a lot unuse code.

That is why I questioned the use of other library

**Comment by jkronegg** [comment]
I agree that readability is generally lower when using stub classes with a lot of methods... This may be improved by having specific stub classes that are shared between test cases, but @mpkorstanje wanted stubs implementations to be declared in the test class. 

I didn't found a lot of mocking libraries performance comparisons, but according to https://github.com/mockito/mockito/issues/3288, mockito and Easymock have more or less the same performance.

**Comment by CharlesLgn** [comment]
> I didn't found a lot of mocking libraries performance comparisons, but according to [mockito/mockito#3288](https://github.com/mockito/mockito/issues/3288), mockito and Easymock have more or less the same performance.

As I understand, this is not what they said.

 https://github.com/mockito/mockito/issues/3288#issue-2165295293

> So you see that EasyMock shows 100x less average execution time. And that's very strange. Can you please check my benchmarks? Have I done something wrong? Or how can you explain that?

the person opening the ticket says that EasyMoch shows 100x less average execution time.

Easy mock
```java
Callable mock;

@Setup
public void setup() throws Exception {
        mock = EasyMock.mock(Callable.class);

	EasyMock.expect(mock.call()).andReturn("1").times(1, Integer.MAX_VALUE);
	EasyMock.replay(mock);
}

@Benchmark
public Object testEastMock() throws Exception {
	return mock.call();
}
```

Mockito
```java
Callable mock;

@Setup
public void setup() throws Exception {
	mock = mock(Callable.class, withSettings().mockMaker(MockMakers.INLINE));
	when(mock.call()).thenReturn("1");
}

@Benchmark
public Object testMockitoInline() throws Exception {
	return mock.call();
}
```

```
Benchmark                              Mode  Cnt   Score     Error    Units
MockitoBenchmarking.testEastMock       avgt    5    46.058 ±   0.443  ns/op
MockitoBenchmarking.testMockitoInline  avgt    5  4428.439 ± 356.722  ns/op
```

but maybe I did not understand what @sergey-morenets is saying 😕

**Comment by mpkorstanje** [comment]
Alright folks take it easy on the bike-shedding. 😉 

As is, I've picked up a few parts of this PR. But overall it is quite boring to review and quite important to get right. So it does not seem to end on top of my priority queue. 

@jkronegg my appologies for that.
